pkg/controllers/node/node_controller.go

```diff
@@ -0,0 +1,294 @@
+			// so there's no other relevant information we want to store in the cache besides the name (which
+			// is unavailable at this time because the calicoNode is created after the k8sNode).
+			k8sResourceCache.Set(nodeName, "")
+		},
+	}, corecache.Indexers{})
+
+	return &NodeController{ informer, k8sResourceCache, &nodeLookupCache, calicoClient, k8sClientset}
+}
+
+// getK8sNodeName is a helper method that searches a calicoNode for its kubernetes nodeRef.
+func getK8sNodeName(calicoNode api.Node) string {
+	for _, orchRef := range calicoNode.Spec.OrchRefs {
+		if orchRef.Orchestrator == "k8s" {
+			return orchRef.NodeName
+		}
+	}
+	return ""
+}
+
+// Run starts the node controller. It does start-of-day preparation
+// and then launches worker threads.
+func (c *NodeController) Run(threadiness int, reconcilerPeriod string, stopCh chan struct{}) {
+	defer uruntime.HandleCrash()
+
+	// Let the workers stop when we are done
+	workqueue := c.k8sResourceCache.GetQueue()
+	defer workqueue.ShutDown()
+
+	log.Info("Starting Node controller")
+
+	// Load node cache. Retry when failed.
+	log.Debug("Loading node cache at start of day")
+	for err := c.populateNodeLookupCache(); err != nil; {
+		log.WithError(err).Errorf("Failed to load Node cache, retrying in 5s")
+		time.Sleep(5 * time.Second)
+	}
+
+	// Wait till k8s cache is synced
+	go c.informer.Run(stopCh)
+	log.Debug("Waiting to sync with Kubernetes API (Nodes)")
+	for !c.informer.HasSynced() {
+	}
+	log.Debug("Finished syncing with Kubernetes API (Nodes)")
+
+	// Start Calico cache.
+	c.k8sResourceCache.Run(reconcilerPeriod)
+
+	// Start a number of worker threads to read from the queue.
+	for i := 0; i < threadiness; i++ {
+		go c.runWorker()
+	}
+	log.Info("Node controller is now running")
+
+	<-stopCh
+	log.Info("Stopping Node controller")
+}
+
+func (c *NodeController) runWorker() {
+	for c.processNextItem() {}
+}
+
+func (c *NodeController) processNextItem() bool {
+	// Wait until there is a new item in the work queue.
+	workqueue := c.k8sResourceCache.GetQueue()
+	key, quit := workqueue.Get()
+	if quit {
+		return false
+	}
+
+	// Sync the object to the Calico datastore.
+	if err := c.syncToCalico(key.(string)); err != nil {
+		c.handleErr(err, key.(string))
+	}
+
+	// Indicate that we're done processing this key, allowing for safe parallel processing such that
+	// two objects with the same key are never processed in parallel.
+	workqueue.Done(key)
+	return true
+}
+
+// populateNodeLookupCache fills the nodeLookupCache with initial data
+// by querying the existing data stored in Calico.
+func (c *NodeController) populateNodeLookupCache() error {
+	nodes, err := c.calicoClient.Nodes().List(api.NodeMetadata{})
+	if err != nil {
+		return err
+	}
+
+	c.nodeLookupCache.Lock()
+	for _, node := range nodes.Items {
+		if k8sNodeName := getK8sNodeName(node); k8sNodeName != "" {
+			c.nodeLookupCache.nodes[k8sNodeName] = nodeData{node.Metadata.Name}
```

Comment:
What's the purpose of the `nodeData` intermediate struct? Should probably just be a string, right?

Reply:
Did I implement this on the wrong cache then? https://github.com/projectcalico/kube-controllers/pull/163#discussion_r143587839